### PR TITLE
Refactoring changes

### PR DESCRIPTION
**Pull request by magnatelee**
This PR contains the following changes:

* Assign unique IDs to `Store` and `Storage` objects
* Move the partition cache from `Store` to `PartitionManger` to avoid mutual references between `Store` and `PartitionBase` that will be introduced in the near future
* Refactor the task slicing logic in `BaseMapper` and `CoreMapper`. this fixes https://github.com/nv-legate/cunumeric/issues/685 as a byproduct.

### CODE REVIEW

**Review by manopapad — APPROVED** [review]
Some small comments

File: legate/core/store.py [hunk 1]
```diff
@@ -886,12 +890,12 @@ def __init__(
         else:
             sanitized_transform = identity
         assert isinstance(shape, Shape) or shape is None
+        self._unique_id = store_id
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/store.py]: https://github.com/nv-legate/legate.core/pull/478/commits/d9dfa44e48becb08286028557b1e2d1a82e360e8

[the remark this answers is not in the document]
Comment (magnatelee) [reply legate/core/store.py]: the id arguments are now removed

File: src/core/mapping/base_mapper.cc [hunk 2]
```diff
@@ -253,23 +254,14 @@ void BaseMapper::slice_auto_task(const MapperContext ctx,
...
-      auto lo = sharding_domain.lo();
-      auto hi = sharding_domain.hi();
-      for (Domain::DomainPointIterator itr(input.domain); itr; itr++) {
-        auto idx = linearize(lo, hi, itr.p);
-        output.slices.push_back(TaskSlice(
-          Domain(itr.p, itr.p), procs[idx % procs.size()], false /*recurse*/, false /*stealable*/));
-      }
+    auto lo     = key_functor->project_point(sharding_domain.lo(), sharding_domain);
+    auto hi     = key_functor->project_point(sharding_domain.hi(), sharding_domain);
+    auto offset = linearize(lo, hi, key_functor->project_point(input.domain.lo(), sharding_domain));
```
[the remark this answers is not in the document]
Comment (magnatelee) [reply src/core/mapping/base_mapper.cc]: https://github.com/nv-legate/legate.core/pull/478/commits/b7ea06932c287270e7b78fdb08f160aa96550141

[the remark this answers is not in the document]
Comment (magnatelee) [reply src/core/mapping/base_mapper.cc]: I ended up just removing offsets for now, and they weren't really necessary to fix the bug in `unique`. I agree that a proper super-tiling is the right way to go and I've actually been contemplating implementing it, but let's defer it to a follow-on PR, as #457 is about to land and that PR could make it easier to implement the super-tiling.

**Review by manopapad — APPROVED** [review]

**Pull request merged**